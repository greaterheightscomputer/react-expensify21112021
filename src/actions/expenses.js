import database from "../firebase/firebase";

//let recap of what we have being doing with Action Generator Function
//1. component calls action generator.
//2. action generator returns an object.
//3. component dispatch action object.
//4. redux store changes.

//this is what we will be doing with Asynchronous Action Generator Function
//1. component calls action generator.
//2. action generator returns function.
//3. component dispatch action function.
//4. function runs (has the ability to change both firebase and redux store)

//Action Generator
//- Action Generator function for expenses array object are as follows;
//ADD_EXPENSE
// export const addExpense = ({
//   description = "",
//   note = "",
//   amount = 0,
//   createdAt = 0,
// } = {}) => ({
//   type: "ADD_EXPENSE",
//   expense: {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt,
//   },
// });
//local Action Generator Function
//ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});
//Asynchronous ADD_EXPENSE Action Generator Function
//- We need to do the following inside startAddExpenese() function
//1. create startAddExpense (same call signature as addExpense)
//2. Test startAddExpense with "should add expense to firebase"
//3. Use startAddExpense in AddExpensePage instead of addExpense action generator
//4. Adjust AddExpensePage test

//asynchronous Action Generator Function
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    //return keyword beside the database means pass the data return in then() promise method to
    //the test case, without it your test case will not be successful.
    return database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        console.log(ref);

        //dispatch local Action Generator function to redux store
        dispatch(
          addExpense({
            id: ref.key, //using ref.key which is the id from firebase in place of uuid()
            ...expense,
          })
        );
      });
  };
};

// //REMOVE_EXPENSE
// export const removeExpense = ({ id } = {}) => ({
//   type: "REMOVE_EXPENSE",
//   id,
// });

//local Remove Expense
//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

//Asynchronous REMOVE_EXPENSE Action Generator Function
//- We need to do the following inside startRemoveExpenese() function
//1. create removeAddExpense (same call signature as removeExpense)
//2. Test removeAddExpense with "should remove expense from firebase"
//3. Use startRemoveExpense in EditExpensePage instead of removeExpense action generator
//4. Adjust EditExpensePage test

export const startRemoveExpenese = ({ id } = {}) => {
  return (dispatch) => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

//EDIT_EXPENSE
// export const editExpense = (id, updates) => ({
//   type: "EDIT_EXPENSE",
//   id,
//   updates,
// });

//local edit expense
//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

//Asynchronous EDIT_EXPENSE Action Generator Function
//- We need to do the following inside startEditExpenese() function
//1. create editAddExpense (same call signature as editExpense)
//2. Test editAddExpense with "should edit expense from firebase"
//3. Use startEditExpense in EditExpensePage instead of editExpense action generator
//4. Adjust EditExpensePage test

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database
      .ref(`expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

//Fetching data from both firebase and redux store
//SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

//Asynchronous EDIT_EXPENSE Action Generator Function
//- We need to do the following inside startSetExpenese() function
//1. fetch all expense data once
//2. parse that data onto an array
//3. dispatch SET_EXPENSE to redux store

export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref("expenses")
      .once("value")
      .then((snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};

//Let add user to the root of ref() to all our Async Action Generator Function inorder to
// change it from .ref("expenses") to .ref(`users/${uid}/expenses`)
//ADD_EXPENSE
// export const addExpense = (expense) => ({
//   type: "ADD_EXPENSE",
//   expense,
// });

// //asynchronous Action Generator Function
// export const startAddExpense = (expenseData = {}) => {
//   //getState() method is to fetch data from redux store
//   return (dispatch, getState) => {
//     const uid = getState().auth.uid; //get the user id from redux store
//     const {
//       description = "",
//       note = "",
//       amount = 0,
//       createdAt = 0,
//     } = expenseData;
//     const expense = { description, note, amount, createdAt };
//     return (
//       database
//         // .ref("expenses")
//         .ref(`users/${uid}/expenses`)
//         .push(expense)
//         .then((ref) => {
//           dispatch(
//             addExpense({
//               id: ref.key,
//               ...expense,
//             })
//           );
//         })
//     );
//   };
// };
// //- let test the app by starting the dev-server then insert data to firebase db to view if its works.
// //- click on Create Expense to add expense
// //- view firebase to see the structure of the current data
// //- you will see the users/7sududisiis89990/expenses/ which is the current data structure for
// // every each individual object will create in our app.

// // //REMOVE_EXPENSE
// // export const removeExpense = ({ id } = {}) => ({
// //   type: "REMOVE_EXPENSE",
// //   id,
// // });

// //local Remove Expense
// //REMOVE_EXPENSE
// export const removeExpense = ({ id } = {}) => ({
//   type: "REMOVE_EXPENSE",
//   id,
// });

// //Asynchronous REMOVE_EXPENSE Action Generator Function
// //- We need to do the following inside startRemoveExpenese() function
// //1. create removeAddExpense (same call signature as removeExpense)
// //2. Test removeAddExpense with "should remove expense from firebase"
// //3. Use startRemoveExpense in EditExpensePage instead of removeExpense action generator
// //4. Adjust EditExpensePage test

// export const startRemoveExpenese = ({ id } = {}) => {
//   return (dispatch) => {
//     return database
//       .ref(`expenses/${id}`)
//       .remove()
//       .then(() => {
//         dispatch(removeExpense({ id }));
//       });
//   };
// };

// //EDIT_EXPENSE
// // export const editExpense = (id, updates) => ({
// //   type: "EDIT_EXPENSE",
// //   id,
// //   updates,
// // });

// //local edit expense
// //EDIT_EXPENSE
// export const editExpense = (id, updates) => ({
//   type: "EDIT_EXPENSE",
//   id,
//   updates,
// });

// //Asynchronous EDIT_EXPENSE Action Generator Function
// //- We need to do the following inside startEditExpenese() function
// //1. create editAddExpense (same call signature as editExpense)
// //2. Test editAddExpense with "should edit expense from firebase"
// //3. Use startEditExpense in EditExpensePage instead of editExpense action generator
// //4. Adjust EditExpensePage test

// export const startEditExpense = (id, updates) => {
//   return (dispatch) => {
//     return database
//       .ref(`expenses/${id}`)
//       .update(updates)
//       .then(() => {
//         dispatch(editExpense(id, updates));
//       });
//   };
// };

// //Fetching data from both firebase and redux store
// //SET_EXPENSES
// export const setExpenses = (expenses) => ({
//   type: "SET_EXPENSES",
//   expenses,
// });

// //Asynchronous EDIT_EXPENSE Action Generator Function
// //- We need to do the following inside startSetExpenese() function
// //1. fetch all expense data once
// //2. parse that data onto an array
// //3. dispatch SET_EXPENSE to redux store

// export const startSetExpenses = () => {
//   return (dispatch) => {
//     return database
//       .ref("expenses")
//       .once("value")
//       .then((snapshot) => {
//         const expenses = [];

//         snapshot.forEach((childSnapshot) => {
//           expenses.push({
//             id: childSnapshot.key,
//             ...childSnapshot.val(),
//           });
//         });
//         dispatch(setExpenses(expenses));
//       });
//   };
// };
