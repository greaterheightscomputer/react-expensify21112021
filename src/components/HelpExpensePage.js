import React from "react";

const HelpExpensePage = () => (
  <div>This is from my help expenses component.</div>
);

export default HelpExpensePage;
