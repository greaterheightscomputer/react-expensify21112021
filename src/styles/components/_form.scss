.form {
  display: flex;
  flex-direction: column;
  > * {
    margin-bottom: $m-size; //>* apply these style to the direct children of form class selector
  }
}
