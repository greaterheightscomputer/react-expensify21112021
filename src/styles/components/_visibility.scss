.show-for-mobile {
  //width <=720px or 45rem display Expenses
  @media (min-width: $desktop-breakpoint) {
    display: none;
  }
}

.show-for-desktop {
  //width >=720px or 45rem display Expense and Amount
  //0.01rem means both max-width and min-width never meet
  @media (max-width: $desktop-breakpoint - 0.01rem) {
    display: none;
  }
}
