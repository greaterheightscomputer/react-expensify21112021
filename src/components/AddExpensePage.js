// import React from "react";
// import { connect } from "react-redux";
// import { addExpense } from "../actions/expenses";
// import ExpenseForm from "./ExpenseForm";

// export class AddExpensePage extends React.Component {
//   onSubmit = (expense) => {
//     this.props.addExpense(expense);
//     this.props.history.push("/");
//   };

//   render() {
//     return (
//       <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm onSubmitForm={this.onSubmit} />
//       </div>
//     );
//   }
// }

// const mapDispatchToProps = (dispatch) => {
//   return {
//     addExpense: (expense) => dispatch(addExpense(expense)),
//   };
// };

// export default connect(undefined, mapDispatchToProps)(AddExpensePage);

//Changing of addExpense() Action Generator Function to startAddExpense() Asynchronous Action
//Generator Function
import React from "react";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmitForm={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
