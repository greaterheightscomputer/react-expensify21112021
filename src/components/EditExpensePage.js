import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpenese } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveExpenese({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmitForm={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

//fetch data from redux store
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

//setup mapDispatchToProps()
const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpenese: (data) => dispatch(startRemoveExpenese(data)),
});
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
