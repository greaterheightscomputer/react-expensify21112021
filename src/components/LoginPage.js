import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

// export const LoginPage = (props) =>  (
//     <div>
//       <button onClick={props.startLogin}>Login</button>
//     </div>
//   );
//

//destructure props and connect LoginPage to redux store
export const LoginPage = ({ startLogin }) => (
  <div>
    <button onClick={startLogin}>Login</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
