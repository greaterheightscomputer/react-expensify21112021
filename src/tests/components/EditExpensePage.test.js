import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

//these are functions we are testing from EditExpensePage component
// onSubmit = (expense) => {
//     this.props.editExpense(this.props.expense.id, expense);
//     this.props.history.push("/");
//   };
//   onRemove = () => {
//     this.props.removeExpense({ id: this.props.expense.id });
//     this.props.history.push("/");
//   };

let editExpense, removeExpense, history, wrapper; //creating variables for editExpense, removeExpense, history to store fake function for props fro EditExpensePage component

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});

test("should render EditExpensePage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense on submit", () => {
  wrapper.find("ExpenseForm").prop("onSubmitForm")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test("should handle removeExpense on removing expense", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});
