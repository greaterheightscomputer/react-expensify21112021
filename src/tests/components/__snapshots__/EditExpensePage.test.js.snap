// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 890,
          "createdAt": -345600000,
          "description": "Shirt",
          "id": "2",
          "note": "",
        }
      }
      onSubmitForm={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
